pytorch_lightning/strategies/launchers/spawn.py

```diff
@@ -0,0 +1,191 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from collections import UserList
+from multiprocessing.queues import SimpleQueue
+from typing import Any, Callable, NamedTuple, Optional
+
+import numpy as np
+import torch
+import torch.multiprocessing as mp
+
+import pytorch_lightning as pl
+from pytorch_lightning.strategies.launchers.base import _Launcher
+from pytorch_lightning.strategies.strategy import Strategy
+from pytorch_lightning.trainer.states import TrainerFn, TrainerState
+from pytorch_lightning.utilities.apply_func import apply_to_collection, move_data_to_device
+from pytorch_lightning.utilities.model_helpers import is_overridden
+from pytorch_lightning.utilities.rank_zero import rank_zero_debug
+from pytorch_lightning.utilities.types import _PATH
+
+
+class _SpawnLauncher(_Launcher):
+    r"""Spawns processes that run a given function in parallel, and joins them all at the end.
+
+    The main process in which this launcher is invoked creates N so-called worker processes (using
+    :func:`torch.multiprocessing.spawn`) that run the given function.
+    Worker processes have a rank that ranges from 0 to N - 1.
+
+    Note:
+        - This launcher requires all objects to be pickleable.
+        - It is important that the entry point to the program/script is guarded by ``if __name__ == "__main__"``.
+
+    Arguments:
+        strategy: A reference to the strategy that is used together with this launcher.
+    """
+
+    def __init__(self, strategy: Strategy) -> None:
+        self._strategy = strategy
+
+    def launch(self, function: Callable, *args: Any, **kwargs: Any) -> Any:
+        """Spawns processes that run the given function in parallel.
+
+        The function is allowed to have a return value. However, when all processes join, only the return value
+        of worker process 0 gets returned from this `launch` method in the main process.
+
+        Arguments:
+            function: The entry point for all spawned processes.
+            *args: Optional positional arguments to be passed to the given function.
+            **kwargs: Optional keyword arguments to be passed to the given function.
+                If a keyword argument named `trainer` is present and is an instance of
+                :class:`~pytorch_lightning.trainer.trainer.Trainer`, a selected set of attributes from the trainer get
+                restored in the main process after processes join. The `trainer` keyword argument will NOT be passed
+                into the function.
+        """
+        trainer = kwargs.pop("trainer", None)
```

Comment:
Why not have the signature as

```py
def launch(self, function: Callable, *args: Any, trainer: Optional[Trainer] = None, **kwargs: Any) -> Any:
```

So you can remove
```suggestion
```

And the kwargs docstring is simplified?

Same comment would apply to the other launcher implementations

Reply:
I'm confused. My proposal should work with Lite and Lightning.

Both are just passing `*args, **kwargs`, the difference is that the "pop" happens with the signature instead of manually inside the method

Reply:
Ah yes, you are right, if it's an optional kwarg then yes